### ISSUE TITLE
Support failed queries

### ISSUE THREAD

**Issue by reuvenpo** [scrtlabs/secret-toolkit#21]
The current implementation of the `snip20` and `snip721` modules fails to parse the query response if it was an error.
This isn't great. The library should know how to handle those cases. This came up while playing around with queries that require a VK, and providing the wrong VK.

**Comment by baedrik** [comment]
@reuvenpo Ah, right, right.  I should probably be able to squeeze this in.  I'm thinking all that needs to be changed so that it forwards the error from the contract instead of displaying a parse error is to change the query functions like so:

```rust
    let answer: TokenInfoResponse =  QueryMsg::TokenInfo {}.query(querier, block_size, callback_code_hash, contract_addr)?;
    Ok(answer.token_info)
```
becomes
```rust
    QueryMsg::TokenInfo {}.query(querier, block_size, callback_code_hash, contract_addr)
```
Does that sound about right?

**Comment by reuvenpo** [comment]
it's a little more nuanced than that, the answer might not be a `TokenInfoResponse` but a [`ViewingKeyError`](https://github.com/enigmampc/snip20-reference-impl/blob/master/src/msg.rs#L398).

**Comment by baedrik** [comment]
Ah, true.  Forgot that snip20s return an Ok(ViewingKeyError), instead of an Err when the viewing key doesn't match

**Comment by baedrik** [comment]
Regarding errors that get returned as Err, the change in the original comment would forward the error message instead of turning it into a parse error, right?  If so those changes should probably be made also

**Comment by baedrik** [comment]
Just a heads up for someone else that might do this if I can't squeeze it in.  Snip721 returns viewing key errors as Err, so the changes will be slightly different for that toolkit

**Comment by baedrik** [comment]
Actually, I just remembered that the toolkit `query` function that the individual functions use has:
```rust
        querier
        .query(&amp;QueryRequest::Wasm(WasmQuery::Smart {
                contract_addr,
                callback_code_hash,
                msg,
            }))
            .map_err(|err| {
                StdError::generic_err(format!("Error performing {} query: {}", self, err))
            })
```
so the contract error should be getting forwarded when it is an Err with the way it is coded now.  So I believe the only change that should be needed is for the snip20 toolkit, to handle any errors that get returned as Ok, which I *believe* is only viewing key errors. Does that sound right @reuvenpo?

**Comment by baedrik** [comment]
@reuvenpo If I end up making the changes, how would you feel if the toolkit turned all Ok(ViewingKeyError) into an Err?  That way the calling contract won't need to pattern match to see whether it can continue processing and can use the `?` instead.  Since it would do a run around on the snip20 decision to have viewing key errors return Ok results, I could understand if you don't want that.  Just my personal thoughts is that if the contract isn't receiving the information it requested and presumably requires to be able to do what it is trying to do, just letting them exit with `?` seems like less work for them.  Even if they want to do special error processing they could still pattern match the Ok and Err.  It would just allow everyone that doesn't need to do additional error processing the ability to use `?`

**Comment by reuvenpo** [comment]
>Ah, true. Forgot that snip20s return an Ok(ViewingKeyError), instead of an Err when the viewing key doesn't match

It's ok lol i completely forgot about that detail 😄 

>Snip721 returns viewing key errors as Err

We should've done the same in SNIP-20... This situation is awkward to support...

>... the only change that should be needed ... [is] to handle any errors that get returned as Ok

yup

>how would you feel if the toolkit turned all Ok(ViewingKeyError)

That would be ideal.

>I could understand if you don't want that

But i do 🙂 

The question is, how do we represent the error?

**Comment by baedrik** [comment]
When picking the viewing key error for snip721, I would have liked to use StdError::Unauthorized, but unfortunately, it doesn't give you a message field to work with, and I do like the wording used for the viewing key error.  So I opted for:
```rust
Err(StdError::generic_err("Wrong viewing key for this address or viewing key not set"))
```
but I'm not attached to it, and could see an argument for a non-descriptive Unauthorized

**Comment by floAr** [comment]
+1 for typed errors.
One of the main gripes frontend wise is that I get a lot of generic errors and have to parse their content to see what kind of error I got. If we have a few specific errors (like, missing viewing keys and only missing viewing keys return StdError::Unauthorized) and list them somewhere that would help by a lot (similar to http status codes)

**Comment by reuvenpo** [comment]
I agree with you @floAr. Maybe what we should've done is include json messages in `StdError::generic_err()`s created in the contracts... It's a good thing to keep in mind for future SNIPs and Secret Contracts. I regret not considering these options before.

That being said, I think we should think about how to support the existing messages returned by the reference implementation. The `ViewingKeyError` message isn't specified at all, but de-facto all the deployed tokens include it. Whether or not the specification should retroactively mention this message type, I think it's fair to say that the toolkit should support it (because it exists in tokens like sSCRT, sETH, SEFI, etc.)

**Comment by reuvenpo** [comment]
We ended up merging a PR by @baedrik (Thanks Baedrik!) that represents this case as an `StdError::Unauthorized`. This is now included in the `v0.2.0` release, also available on crates.io 😄
